src/main/kotlin/org/codefreak/codefreak/service/file/FileSystemFileService.kt

```diff
@@ -0,0 +1,375 @@
+      require(!isBlacklistedPath(collectionId, path)) { "`$it` is not a valid path" }
+      require(Files.exists(path)) { "`$it` does not exist" }
+      path
+    }
+
+    sanitizedPaths.forEach {
+      FileUtils.forceDelete(it.toFile())
+    }
+  }
+
+  override fun renameFile(collectionId: UUID, source: String, target: String) {
+    val collectionPath = createCollectionPath(collectionId).toString()
+    val sourcePath = Paths.get(collectionPath, source)
+    val targetPath = Paths.get(collectionPath, target)
+
+    require(Files.exists(sourcePath)) { "The source path `$source` does not exist" }
+    require(!isBlacklistedPath(collectionId, sourcePath)) { "The source path `$source` is not a valid path" }
+
+    if (arePathsEqual(sourcePath, targetPath)) {
+      // Renaming file to itself, ignore
+      return
+    }
+
+    require(!Files.exists(targetPath)) { "The target path `$target` already exists" }
+    require(!isBlacklistedPath(collectionId, targetPath)) { "The target path `$target` is not a valid path" }
+
+    val sourceFile = sourcePath.toFile()
+    val targetFile = targetPath.toFile()
+
+    when {
+      Files.isRegularFile(sourcePath) -> FileUtils.moveFile(sourceFile, targetFile)
+      Files.isDirectory(sourcePath) -> FileUtils.moveDirectory(sourceFile, targetFile)
+      else -> throw IllegalStateException("`$sourcePath` does not exist though it should at this point")
+    }
+  }
+
+  override fun moveFile(collectionId: UUID, sources: Set<String>, target: String) {
+    require(sources.isNotEmpty()) { "No sources given" }
+
+    val collectionPath = createCollectionPath(collectionId).toString()
+    val targetPath = Paths.get(collectionPath, target)
+
+    val sourcePaths = mutableMapOf<String, Path>()
+    sources.forEach {
+      val sanitizedPath = FileUtil.sanitizeName(it)
+      val sourcePath = Paths.get(collectionPath, sanitizedPath)
+      require(Files.exists(sourcePath)) { "The source path `$it` does not exist" }
+      require(!isBlacklistedPath(collectionId, sourcePath)) { "The source path `$it` is not a valid path" }
+
+      val mappedTargetPath = Paths.get(targetPath.toString(), sanitizedPath)
+      val shouldMovePath = !arePathsEqual(sourcePath, targetPath) && !arePathsEqual(sourcePath, mappedTargetPath) || Files.isDirectory(sourcePath)
+
+      if (shouldMovePath) {
+        sourcePaths[it] = sourcePath
+      }
+    }
+
+    if (sourcePaths.isEmpty()) {
+      // Nothing to move
+      return
+    }
+
+    require(Files.isDirectory(targetPath)) { "`$target` is not a directory" }
+    require(!isBlacklistedPath(collectionId, targetPath)) { "The target path `$target` is not a valid path" }
+
+    sourcePaths.forEach { (originalPath, sourcePath) ->
+      require(!isDescendant(targetPath, sourcePath)) { "The target `$target` is a descendant of the source `$originalPath`" }
+      require(!isBasenamePresent(sourcePath, targetPath)) { "`$originalPath` or a child of it already exists in `$targetPath`" }
```

Comment:
`FileUtils.moveToDirectory` will throw exceptions for both of this cases

Reply:
Yes - but our current behaviour (through the tests and the JpaFileService) is that nothing is moved if there is an invalid file. So I did the checks before moving any files. Should we change the behaviour that it already moves files until there is a file that does not meet those conditions?

Reply:
Yeah just try to move it. classic "ask forgiveness not permission"